### PR TITLE
feat: GPU shared texture offscreen rendering

### PR DESCRIPTION
**Pull request by reitowo**
#### Description of Change

Reopen #42001 

I managed to add `ARGB + kGpuMemoryBuffer` support for FrameSinkVideoCapturer in [recent upstream changes](https://chromiumdash.appspot.com/commits?user=carolwolfking&platform=Windows). Thus, we can finally use zero-copy (actually one, there's a CopyRequest of frame texture) GPU shared texture OSR in chromium apps.

This will be the fastest OSR mode than the existing two, and it supports enabling hardware acceleration.

> However, this mode will not compose the popup widgets, directly provide both window types' textures to user. It is expected that user themselves to compose these textures at their rendering pipeline.

Details:
1. Add `webPreferences.offscreen.useSharedTexture` to enable this feature.
2. Add `texture` to `paint` event of `webContent`.
3. Add structure definitions and docs.
4. Add tests about this feature.

Fix: #41972

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [x] relevant documentation, tutorials, templates and examples are changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/main/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/main/README.md#examples).

#### Release Notes

Notes: feat: GPU accelerated shared texture offscreen rendering.

### CODE REVIEW

**Comment by reitowo** [comment]
Well, it seems the CI device failed to create a D3D11 device. Anything I can do about this?

image

File: docs/api/web-contents.md
Comment (samuelmaddock) [inline docs/api/web-contents.md:898]: I think it might make sense to split up these two examples. First the existing example, then one for when `offscreenUseSharedTexture` is enabled. Most applications will choose either one or the other so having the code separate might be more intuitive.

File: docs/tutorial/offscreen-rendering.md
Comment (samuelmaddock) [inline docs/tutorial/offscreen-rendering.md:30]: Simpler for non-native speakers
```suggestion
GPU accelerated rendering means that the GPU is used for composition. The benefit
of this mode is that WebGL and 3D CSS animations are supported. There are two
different approaches depending on whether `webPreferences.offscreenUseSharedTexture`
is set to true.
```

File: docs/tutorial/offscreen-rendering.md
Comment (samuelmaddock) [inline docs/tutorial/offscreen-rendering.md:43]: For (1), I think it should be mentioned that this is an advanced approach which requires a Node native module to handle the texture. For (2), it should mention that the texture is accessible using the `NativeImage` JavaScript API at the cost of performance.

File: shell/browser/api/electron_api_web_contents.cc
Comment (samuelmaddock) [inline shell/browser/api/electron_api_web_contents.cc:3543]: Only set if enabled maybe? This would align with the docs about it being optional.
```suggestion
  if (offscreen_use_shared_texture_)
      dict.Set("texture", tex);
```

File: spec/fixtures/native-addon/osr-gpu/binding_win.cc
Comment (samuelmaddock) [inline spec/fixtures/native-addon/osr-gpu/binding_win.cc:197]: Nicely done with this everything involving this test 👏

File: patches/chromium/osr_shared_texture_remove_keyed_mutex_on_win_dxgi.patch
Comment (samuelmaddock) [inline patches/chromium/osr_shared_texture_remove_keyed_mutex_on_win_dxgi.patch:64]: Electron's desktopCapturer uses this format. Would it be affected?

https://github.com/electron/electron/blob/7a101a1c4b104b3844ab12b66980246df42f8560/shell/browser/api/frame_subscriber.cc#L50

Comment (reitowo) [reply patches/chromium/osr_shared_texture_remove_keyed_mutex_on_win_dxgi.patch:64]: As long as this code use `kDefault` which requests a CPU only buffer (https://github.com/electron/electron/blob/7a101a1c4b104b3844ab12b66980246df42f8560/shell/browser/api/frame_subscriber.cc#L52C4-L52C78), it won't be affected by this. Even it use GPU texture, it will not be affected if you won't write to the texture again. The patch just remove the keyed mutex on that texture.

**Comment by reitowo** [comment]
Rebased conflict.

**Comment by reitowo** [comment]
Any updates?

File: docs/api/structures/web-preferences.md
Comment (samuelmaddock) [inline docs/api/structures/web-preferences.md:86]: Since we haven't introduced new options for offscreen rendering prior to this PR, we're thinking of changing the `offscreen` property itself to accept an object or boolean.

* `offscreen` Object | Boolean (optional)
   * `useSharedTexture` boolean (optional) _Experimental_ - Whether to use GPU shared texture for accelerated color.

This would allow for future options to more easily be added.

We have precedence for this with the `titleBarOverlay` options if you're looking for a code reference.

https://github.com/electron/electron/blob/5b18cc46bc4ef293a5cfe54c52b94447bd7dc11c/shell/browser/native_window.cc#L111-L125

File: docs/api/web-contents.md
Comment (erickzhao) [inline docs/api/web-contents.md:892]: I'm unfamiliar with the underlying concepts here. Is there a general rule of thumb on the "before the underlying frame pool is drained" part?

And would it be possible to explain what happens if we don't release the texture before the frame pool is drained? (e.g. Will the app crash?)

File: docs/api/structures/web-preferences.md
Comment (reitowo) [reply docs/api/structures/web-preferences.md:86]: Good idea!

File: docs/api/structures/offscreen-shared-texture.md
Comment (erickzhao) [inline docs/api/structures/offscreen-shared-texture.md:1]: I think we'd need to see if our Docs Parser code supports this, but I wonder if it'd be possible to mark this Object as experimental as well since it's only used in conjunction with experimental APIs.

**Review by erickzhao — COMMENTED** [review]
Thanks for this PR! Some API comments. :)

File: docs/api/structures/offscreen-shared-texture.md
Comment (erickzhao) [reply docs/api/structures/offscreen-shared-texture.md:1]: We could just test out the syntax and see if the Artifact Comparison violently fails or adds the proper `@experimental` TSDoc tag: https://github.com/electron/electron/pull/42953/checks?check_run_id=28222213375

File: docs/api/web-contents.md
Comment (reitowo) [reply docs/api/web-contents.md:892]: The underlying chromium use FrameSinkVideoCapturer for this feature, and it use RenderableGpuMemoryBufferVideoFramePool to allocate frames. Last time I check it has a const size around 10.

If the buffer is drained the app won't crash, but no more paint events will be populated, unless release is called.

File: docs/api/structures/offscreen-shared-texture.md
Comment (reitowo) [reply docs/api/structures/offscreen-shared-texture.md:1]: I checked the generated electron.d.ts and it seems good? Ok, I'll add experimental.

**Comment by itsananderson** [comment]
If someone enables this feature, but neglects to call `texture.release()`, it sounds like that could cause a memory leak. Would it be possible to monitor for when the JS `texture` object is getting GC'd and either release the shared texture automatically (if that is safe to do), or print a warning so that the developer can detect that they're leaking memory?

There is an [EmitWarning ](https://github.com/electron/electron/blob/main/shell/common/process_util.cc#L16-L28) function that you could use to log this warning. You'd also only want to log it once per process lifetime, to avoid spamming for every `paint` event.

**Comment by reitowo** [comment]
Definetely ok, but not necessary. People use OSR features are expected to receive continuous frames. And since it will be drained only after 10-11 frames, the user will no longer receive any updates, so they must notice this as a bug.

Relying on GC is not working because the unknown timing. 10 frames are only 160ms at 60fps and GC are not running that fast.

However I think it is best to add that hint when it got GC-ed but not manually released for better understanding whats going on. 👍🏻

File: docs/api/structures/offscreen-shared-texture.md
Comment (reitowo) [reply docs/api/structures/offscreen-shared-texture.md:1]: Seems not able to add  _Experimental_ behind or before `Object`.

**Comment by reitowo** [comment]
Implemented it, fun feature.
image

**Comment by samuelmaddock** [comment]
@reitowo looks like the patch needs to be updated as it's failing to apply.

**Comment by reitowo** [comment]
Fixed the patch and verified to work! Hope we can merge this in next days 😎

**Comment by reitowo** [comment]
Guys, this PR is kind of big, and enters merge conflict frequently. Would you spare some time review this soon? ❤️

File: patches/chromium/osr_shared_texture_remove_keyed_mutex_on_win_dxgi.patch
Comment (jkleinsc) [inline patches/chromium/osr_shared_texture_remove_keyed_mutex_on_win_dxgi.patch:15]: This description is not helpful as it is currently written.  Please update the description to explain exactly what this patch accomplishes.

File: docs/tutorial/offscreen-rendering.md
Comment (jkleinsc) [inline docs/tutorial/offscreen-rendering.md:30]:
```suggestion
different approaches depending on the `webPreferences.offscreen.useSharedTexture`
setting.
```

File: docs/tutorial/offscreen-rendering.md
Comment (jkleinsc) [inline docs/tutorial/offscreen-rendering.md:33]:
```suggestion

Used when `webPreferences.offscreen.useSharedTexture` is set to `true`.

```

File: docs/tutorial/offscreen-rendering.md
Comment (jkleinsc) [inline docs/tutorial/offscreen-rendering.md:40]:
```suggestion

Used when `webPreferences.offscreen.useSharedTexture` is set to `false` (default behavior).

```

File: docs/tutorial/offscreen-rendering.md
Comment (jkleinsc) [inline docs/tutorial/offscreen-rendering.md:30]: This section should explicitly call out what value of `webPreferences.offscreen.useSharedTexture` applies to which approach.  I've added a suggestion on how to do this, but feel free to word it differently.

**Review by jkleinsc — CHANGES_REQUESTED** [review]
Overall this looks good, but the documentation and patch description need updating.

**Comment by reitowo** [comment]
Done.

**Review by samuelmaddock — APPROVED** [review]
API LGTM

**Review by jkleinsc — COMMENTED** [review]
API LGTM

**Comment by reitowo** [comment]
🎉 Thanks!

**Comment by reitowo** [comment]
@erickzhao @ckerr Guys, could you please recheck this and approve once you got time, thanks.
@jkleinsc It is necessary for you to approve the changes to unblock, unsure about that 😜, just to make sure

**Review by jkleinsc — APPROVED** [review]

File: docs/api/structures/offscreen-shared-texture.md
Comment (erickzhao) [reply docs/api/structures/offscreen-shared-texture.md:1]: That seems fine to me :)

**Review by erickzhao — APPROVED** [review]
API LGTM

File: shell/common/gin_converters/osr_converter.cc
Comment (ckerr) [inline shell/common/gin_converters/osr_converter.cc:129]: Looks like `plane` on 122 is an accidental copy rather than a reference. Maybe

```c++
  for (const auto& plane : val.planes) {
```

or even fancier,

```suggestion
  auto v8_planes = base::ToVector(val.planes, [&isolate](const auto& plane){
    gin::Dictionary v8_plane(isolate, v8::Object::New(isolate));
    v8_plane.Set("stride", plane.stride);
    v8_plane.Set("offset", plane.offset);
    v8_plane.Set("size", plane.size);
    v8_plane.Set("fd", plane.fd);
    return v8_plane;
  });
```

(will also need `#include "base/containers/to_vector.h"` at top)

File: shell/browser/osr/osr_video_consumer.cc
Comment (ckerr) [inline shell/browser/osr/osr_video_consumer.cc:35]: I don't understand this comment about preventing delay.

File: shell/common/gin_converters/osr_converter.cc
Comment (ckerr) [inline shell/common/gin_converters/osr_converter.cc:146]: Nice work here :+1:

File: docs/api/structures/offscreen-shared-texture.md
Comment (ckerr) [inline docs/api/structures/offscreen-shared-texture.md:12]:
```suggestion
    * `regionCaptureRect` [Rectangle](rectangle.md) (optional) - May reflect the frame's contents origin if region capture is used internally.
```

File: docs/api/web-contents.md
Comment (ckerr) [inline docs/api/web-contents.md:892]:
```suggestion
Only a limited number of textures can exist at the same time, so it's important that you call `texure.release()` as soon as you're done with the texture.
```

File: docs/api/structures/offscreen-shared-texture.md
Comment (ckerr) [inline docs/api/structures/offscreen-shared-texture.md:22]: I'd include the "important to call this as soon as you're done with the texture" warning here too

**Review by ckerr — COMMENTED** [review]
Overall this is a really nice patch @reitowo. I like the tests and especially like the attention given to supporting macOS, Linux, and Windows altogether :+1:

**Comment by ckerr** [comment]
Also CC @miniak who may be interested

File: shell/browser/osr/osr_video_consumer.cc
Comment (reitowo) [reply shell/browser/osr/osr_video_consumer.cc:35]: Previously when the view's size changed (the windows dragged, maximize, etc..) it will set the new size in the `SizeChanged` and also, `SetResolutionConstraints` to limit the video capturer limit the exactly size to the view's size. 

However, there's a bug because of that. When the windows's size changes fastly, the video capturer will give out unexpected size and padding textures. It even resize the frame because of that, causing the rendered frame to be blurry. The reason is the **delay** introduced by this constraint, because the video capturer may not use the constraint until next frame (or maybe more), it may resize the frame to fit the old constraint which is not the expected behavior (lossless frames).

Then we figure out that after removing this constraint, it will still give out correct frames and doesn't affect the behavior. Actually, the resolution constriant shouldn't be used like it used to be, it don't work well with frequent changes, it should be used when you want to limit to 1080P for whole capture lifetime.

Comment (reitowo) [reply shell/browser/osr/osr_video_consumer.cc:35]: Comments also updated :)

**Comment by reitowo** [comment]
Made changes according to @ckerr suggestions! Thanks! ❤️

**Comment by reitowo** [comment]
The CI seems got a flaky linux test, and WoA machine is not fast enough to complete build before 3hrs limit.

**Comment by reitowo** [comment]
Good news, it passed tests.

**Comment by ckerr** [comment]
My concern is that if we merge, then all PRs to `main` will start failing to build too.

I'm confused. Yeah this is a big PR, but it's not _that_ big and other PRs are going green on WoA. What's causing this PR to be an outlier?

@jkleinsc any idea what's going on? ^

**Comment by ckerr** [comment]
> I think there's no point to retry WoA build test unless upgrade the compiling speed... It always fail after 3hrs timeout. We have retried multiple times. It stops around 65000/82000, so probably 5hrs-6hrs is enough for it.

Looks like whatever was going on with WoA may have cleared up? https://ci.appveyor.com/project/electron-bot/electron-woa-testing/builds/50456366 is showing it built in 1 hr 39 min and tests are running now.

**Comment by reitowo** [comment]
> Looks like whatever was going on with WoA may have cleared up? https://ci.appveyor.com/project/electron-bot/electron-woa-testing/builds/50456366 is showing it built in 1 hr 39 min and tests are running now.

wow, glad to hear it comes back then.

**Comment by ckerr** [comment]
merging this at last: everything is green and this has four approvals :tada: 

@reitowo it's uncommon for the project to get fully-formed PRs like this one. Don't be a stranger.

**Comment by reitowo** [comment]
Great!

Thanks for all your work reviewing this PR and maintaining electron. I'm glad to be able to make this contribution! ❤️

**Comment by reitowo** [comment]
Just to ask, is it possible to backport this feature to previous Electron versions?

**Comment by erickzhao** [comment]
Hi @reitowo,

If you would like to backport the PR to previous branches, a releaser can assist by running Trop automation. Backporting a feature PR will automatically put it up for a vote by the Releases working group, whereafter the PR can be accepted or denied on a case-by-case basis.

See https://www.electronjs.org/docs/latest/tutorial/electron-versioning#backport-request-process for more info.

As an aside (I'm not in the WG), I think the general policy for stable release lines is that PRs with large patches or changes to existing code paths are considered riskier and have a lower chance of being accepted.

**Comment by VerteDinde** [comment]
Thanks @erickzhao! @reitowo, I'll go ahead and run backports for you to 33-x-y and 32-x-y. To Erick's point, I think we're probably unlikely to backport a change this large into an already existing stable release line, but we can vote and determine whether or not it's a good fit 🙂

**Comment by reitowo** [comment]
Thanks! I got your point about bp into 32, thats reasonable, and it will still be nice if we can bp into 33 😉.

**Comment by continue-100** [comment]
I know this is being slightly rude, but I'm afraid my [question](https://stackoverflow.com/q/79647606/13353975) would be resigned eternally to the depths of SO, hence, I'm reopening this issue.

**Pull request merged**
